src/BottomUp.c

```diff
@@ -108,6 +118,99 @@ struct UserArgs {
     #endif
 };
 
+/*
+ * new_pathname() -
+ *   Allocates and initializes a new pathname from the given dirname and basename.
+ *
+ *     - dirname_len and basename_len should be WITHOUT the null terminator.
+ *
+ *     - If basename is NULL, then this only uses the dirname and doesn't append a path component
+ *       for the basename.
+ *
+ *   Fills in the path and path length in the given struct BottomUp.
+ */
+static void new_pathname(struct BottomUp *work, const char *dirname, size_t dirname_len,
+                         const char *basename, size_t basename_len) {
+    size_t new_len = dirname_len; // does NOT count null terminator
+
+    if (basename) {
+        new_len += basename_len + 1; // +1 for path separator
+    }
+
+    char *path = calloc(new_len + 1, sizeof(char));  // Additional +1 for null terminator
+    if (!path) {
+        fprintf(stderr, "%s(): out of memory", __func__);
+        /* No point in continuing if we run out of memory: */
+        exit(1);
+    }
+
+    if (basename)
+        SNFORMAT_S(path, new_len + 1, 3, dirname, dirname_len, "/", (size_t) 1, basename, basename_len);
+    else
+        strncpy(path, dirname, new_len + 1);
+
+    work->name = path;
+    work->name_len = new_len;
+}
+
+/*
+ * new_alt_pathname() -
+ *   Allocates and initializes a new alternate pathname, that is, a name that escapes
+ *   characters that cannot appear in a sqlite uri.
+ *
+ *   Parameters [dir|base]name[_len] are the same as in new_pathname().
+ *
+ *   Fills in alt_name and alt_name_len in the given struct BottomUp.
+ *
+ *   Returns 0 if conversion succeeded and 1 if conversion failed, because the output
+ *   buffer wasn't large enough to hold the new name.
+ */
+static int new_alt_pathname(struct BottomUp *work, const char *dirname, size_t dirname_len,
+                             const char *basename, size_t basename_len) {
+    // Alternate name may be longer than the provided dirname and basename together
+    char buf[MAXPATH] = { 0 };
+    size_t bufsize = sizeof(buf);
+    size_t new_len = 0;
+
+    if (basename) {
+        // Assume dirname is already sanitized, only need to sanitize basename:
+        new_len = SNFORMAT_S(buf, bufsize, 2, dirname, dirname_len, "/", (size_t) 1);
+
+        size_t converted_len = basename_len;
+        new_len += sqlite_uri_path(buf + new_len, bufsize - new_len, basename, &converted_len);
+
+        if (converted_len < basename_len) {
+            // uh-oh ... didn't convert entire string
+            return -1;
+        }
+    } else {
+        // Need to sanitize dirname:
+        size_t converted_len = dirname_len;
+        new_len = sqlite_uri_path(buf, bufsize, dirname, &converted_len);
+
+        if (converted_len < dirname_len) {
+            // uh-oh ... didn't convert entire string
+            return -1;
+        }
+    }
+
+    char *path = calloc(new_len + 1, sizeof(char));
+    if (!path)
+        goto oom;
```

Comment:
style: always use `{}` in if blocks, even single lines

Reply:
I just got rid of the `goto` altogether since now there's only one error path in this function.